polkadot/node/tracking-allocator/src/lib.rs

```diff
@@ -0,0 +1,181 @@
+				.is_ok()
+			{
+				break
+			}
+			// We failed to acquire the lock; wait until it's unlocked.
+			//
+			// In theory this should result in less coherency traffic as unlike `compare_exchange`
+			// it is a read-only operation, so multiple cores can execute it simultaneously
+			// without taking an exclusive lock over the cache line.
+			while self.lock.load(Ordering::Relaxed) {
+				std::hint::spin_loop();
+			}
+		}
+	}
+
+	#[inline]
+	fn unlock(&self) {
+		self.lock.store(false, Ordering::Release);
+	}
+
+	fn start_tracking(&mut self, limit: isize, failure_handler: Option<Box<dyn Fn()>>) {
+		self.lock();
+		self.current = 0;
+		self.peak = 0;
+		self.limit = limit;
+		// Cannot drop it yet, as it would trigger a deallocation
+		let old_handler = self.failure_handler.take();
+		self.failure_handler = failure_handler;
+		self.unlock();
+		core::mem::drop(old_handler);
+	}
+
+	fn end_tracking(&mut self) -> isize {
+		self.lock();
+		let peak = self.peak;
+		self.limit = 0;
+		// Cannot drop it yet, as it would trigger a deallocation
+		let old_handler = self.failure_handler.take();
+		self.unlock();
+		core::mem::drop(old_handler);
+		peak
+	}
+
+	#[inline]
+	fn track(&mut self, alloc: isize) -> bool {
+		self.lock();
+		self.current += alloc;
+		if self.current > self.peak {
+			self.peak = self.current;
+		}
+		let within_limits = self.limit == 0 || self.peak <= self.limit;
+		if within_limits {
+			self.unlock()
+		}
+		within_limits
+	}
+}
+
+static mut ALLOCATOR_DATA: TrackingAllocatorData = TrackingAllocatorData {
+	lock: AtomicBool::new(false),
+	current: 0,
+	peak: 0,
+	limit: 0,
+	failure_handler: None,
+};
+
+pub struct TrackingAllocator<A: GlobalAlloc>(pub A);
+
+impl<A: GlobalAlloc> TrackingAllocator<A> {
+	// SAFETY:
+	// * The following functions write to `static mut`. That is safe as the critical section inside
+	//   is isolated by an exclusive lock.
+
+	/// Start tracking
+	/// SAFETY: Failure handler is called with the allocator being in the locked state. Thus, no
+	/// allocations or deallocations are allowed inside the failure handler; otherwise, a
+	/// deadlock will occur.
+	pub unsafe fn start_tracking(
+		&self,
+		limit: Option<isize>,
+		failure_handler: Option<Box<dyn Fn()>>,
+	) {
+		ALLOCATOR_DATA.start_tracking(limit.unwrap_or(0), failure_handler);
+	}
+
+	/// End tracking and return the peak allocation value in bytes (as `isize`). Peak allocation
+	/// value is not guaranteed to be neither non-zero nor positive.
+	pub fn end_tracking(&self) -> isize {
+		unsafe { ALLOCATOR_DATA.end_tracking() }
+	}
+}
+
+#[cold]
+#[inline(never)]
+unsafe fn fail_allocation() -> *mut u8 {
+	if let Some(failure_handler) = &ALLOCATOR_DATA.failure_handler {
+		failure_handler()
+	}
+	ALLOCATOR_DATA.unlock();
+	null_mut()
+}
+
+unsafe impl<A: GlobalAlloc> GlobalAlloc for TrackingAllocator<A> {
+	// SAFETY:
+	// * The wrapped methods are as safe as the underlying allocator implementation is
```

Comment:
If the underlying implementation is safe, why is the `unsafe` then even needed?

Reply:
Unsafety is defined by the trait: https://doc.rust-lang.org/stable/std/alloc/trait.GlobalAlloc.html

Reply:
:bulb: Got it.